File: app/src/main/java/com/johnpetitto/orachat/ui/account/AccountPresenter.java
package com.johnpetitto.orachat.ui.account;

import com.johnpetitto.orachat.data.user.UserModel;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;

public class AccountPresenter {
    private AccountView view;
    private UserModel model;
    private Disposable disposable;

    public AccountPresenter(AccountView view, UserModel model) {
        this.view = view;
        this.model = model;
    }

    public void onDestroy() {
        if (disposable != null) {
            disposable.dispose();
        }
    }

    public void getCurrentUserInfo() {
        view.showLoading(true);

        disposable = model.getCurrentUser()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(user -> {
                    view.populateAccountInfo(user.getName(), user.getEmail());
                    view.showLoading(false);
                }, throwable -> {
                    view.showError();
                    view.showLoading(false);
                });
    }

    public void updateCurrentUser(String name, String email) {
        view.showLoading(true);

        model.updateCurrentUser(name, email)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(() -> view.showLoading(false), throwable -> {
                    view.showError();
                    view.showLoading(false);
                });
    }

    public void logout() {
        model.logout().onErrorComplete().subscribe();
    }
}

File: app/src/main/java/com/johnpetitto/orachat/ui/account/AccountPresenter.kt
package com.johnpetitto.orachat.ui.account

import com.johnpetitto.orachat.data.user.UserModel
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable

class AccountPresenter(private val view: AccountView, private val model: UserModel) {
    private var disposable: Disposable? = null

    fun destroy() = disposable?.dispose()

    fun getCurrentUserInfo() {
        view.showLoading(true)

        disposable = model.getCurrentUser()
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate { view.showLoading(false) }
                .subscribe(
                        { (_, name, email) -> view.populateAccountInfo(name, email) },
                        { view.showError() }
                )
    }

    fun updateCurrentUser(name: String, email: String) {
        view.showLoading(true)

        disposable = model.updateCurrentUser(name, email)
                .observeOn(AndroidSchedulers.mainThread())
                .doAfterTerminate { view.showLoading(false) }
                .subscribe(
                        { view.populateAccountInfo(name, email) },
                        { view.showError() }
                )
    }

    fun logout() = model.logout().onErrorComplete().subscribe()
}
